File: 02_Sorting_Searching/10_Merge_Sort/base_code.py
def merge_sort(arr):
    """
    병합 정렬의 시작점. 임시 배열을 한 번만 할당하여 메모리 낭비를 막습니다.
    """
    temp = [0] * len(arr)
    _merge_sort_recursive(arr, temp, 0, len(arr) - 1)

def _merge_sort_recursive(arr, temp, left, right):
    if left < right:
        mid = (left + right) // 2
        
        # 1. 반으로 나누어 각각 정렬 (Divide)
        _merge_sort_recursive(arr, temp, left, mid)
        _merge_sort_recursive(arr, temp, mid + 1, right)
        
        # 2. 정렬된 두 부분 배열을 병합 (Merge)
        _merge(arr, temp, left, mid, right)

def _merge(arr, temp, left, mid, right):
    for i in range(left, right + 1):
        temp[i] = arr[i]
        
    i = left      # 왼쪽 부분 배열의 포인터
    j = mid + 1   # 오른쪽 부분 배열의 포인터
    k = left      # 원본 배열에 채워 넣을 위치 포인터

    # 두 배열을 비교하며 더 작은 값을 원본 배열에 복사
    while i <= mid and j <= right:
        if temp[i] <= temp[j]:  # <= 를 사용하여 안정 정렬(Stable Sort) 유지
            arr[k] = temp[i]
            i += 1
        else:
            arr[k] = temp[j]
            j += 1
        k += 1

    # 왼쪽 배열에 남은 원소들이 있다면 마저 복사
    # (오른쪽 배열은 이미 원본 배열의 끝자리에 있으므로 처리할 필요 없음)
    while i <= mid:
        arr[k] = temp[i]
        k += 1
        i += 1

# 사용 예시
if __name__ == "__main__":
    data = [38, 27, 43, 3, 9, 82, 10]
    print("정렬 전:", data)
    merge_sort(data)
    print("정렬 후:", data)

File: 02_Sorting_Searching/10_Merge_Sort/BaseCode.java
import java.util.Arrays;

public class BaseCode {
    
    /**
     * 병합 정렬의 시작점. 임시 배열을 한 번만 생성하여 넘겨줍니다.
     */
    public static void mergeSort(int[] arr) {
        int[] temp = new int[arr.length];
        mergeSortRecursive(arr, temp, 0, arr.length - 1);
    }

    private static void mergeSortRecursive(int[] arr, int[] temp, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2; // (left + right) / 2 와 같지만 오버플로우 방지
            
            // 1. 반으로 나누어 각각 정렬 (Divide)
            mergeSortRecursive(arr, temp, left, mid);
            mergeSortRecursive(arr, temp, mid + 1, right);
            
            // 2. 정렬된 두 부분 배열을 병합 (Merge)
            merge(arr, temp, left, mid, right);
        }
    }

    private static void merge(int[] arr, int[] temp, int left, int mid, int right) {
        // 임시 배열에 현재 범위의 데이터 복사
        for (int i = left; i <= right; i++) {
            temp[i] = arr[i];
        }

        int i = left;      // 왼쪽 부분 배열의 포인터
        int j = mid + 1;   // 오른쪽 부분 배열의 포인터
        int k = left;      // 원본 배열에 덮어쓸 위치 포인터

        // 두 부분 배열을 비교하여 더 작은 값을 원본 배열에 채워 넣음
        while (i <= mid && j <= right) {
            if (temp[i] <= temp[j]) { // 안정 정렬 유지
                arr[k] = temp[i];
                i++;
            } else {
                arr[k] = temp[j];
                j++;
            }
            k++;
        }

        // 왼쪽 배열에 남은 원소가 있다면 일괄 복사
        while (i <= mid) {
            arr[k] = temp[i];
            k++;
            i++;
        }
    }

    // 사용 예시
    public static void main(String[] args) {
        int[] data = {38, 27, 43, 3, 9, 82, 10};
        System.out.println("정렬 전: " + Arrays.toString(data));
        mergeSort(data);
        System.out.println("정렬 후: " + Arrays.toString(data));
    }
}